### PR TITLE
RavenDB-16614 Atomic read write in cluster wide transaction without c…

### PR DESCRIPTION
**Pull request by ayende**
…onflicts.

Not done yet - allowing the user to change from the default.
Properly testing deletes.

### CODE REVIEW

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs [hunk 1]
```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
...
+                            }
+
+                            ClusterCommands.Add(new ClusterTransactionDataCommand
+                            {
+                                Type = CommandType.CompareExchangePUT,
+                                Id = atomicGuardKey,
+                                Index = putIndex,
+                                Document = context.ReadObject(new DynamicJsonValue {["Id"] = docId}, "cmp-xchg-content")
+                            });
+                            if (metadata != null) // just to be on the safe side, should never happen
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs]: why this should never happen ?

Comment (karmeli87) [reply src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs]: What would happened if the `metadata` is `null`, should we create a new `metadata`?

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs [hunk 2]
```diff
@@ -66,6 +66,7 @@ public async Task BulkDocs()
                     BatchTrafficWatch(command.ParsedCommands);
                 }
 
+                var disableAtomicDocumentWrites = GetBoolValueQueryString("disableAtomicDocumentWrites", required: false) ?? false;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Handlers/BatchHandler.cs]: How does the user set the `disableAtomicDocumentWrites` option?
to maintain compatibility, maybe we need to allow setting this globally on the `DocumentStore`?

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs [hunk 3]
```diff
@@ -190,15 +200,73 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
...
+
+                            break;
+                        case nameof(CommandType.DELETE):
+                            if (document.TryGetWithoutThrowingOnError(Constants.Documents.Metadata.ClusterTransactionIndex,
+                                out long deleteIndex))
+                            {
+                                // this can happen if the user tried to delete without first loading the document
+                                // we can safely assume that such a blind write doesn't care about the state of the document
+                                // and just proceed normally
+                                deleteIndex = -1;
```
Comment (karmeli87) [inline src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs]: This is actually a problem.. I think.

If I delete and modify concurrently I might get successful response for both, while the document might or might not exist and we back to the original issue.

I think it would be better to enforce loading the document first (check on the client side as well?).

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs]: If this happens, that means that the user submitted a document without metadata.
The only way that can happen is if they crafted a special request. I'm including this check here to avoid NRE in that case, but this isn't something likely.

File: src/Raven.Server/Documents/Handlers/BatchHandler.cs (hunk 2 above)
Comment (ayende) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs]: Currently, they don't, haven't gotten to it

Comment (ayende) [reply src/Raven.Server/Documents/Handlers/BatchHandler.cs]: And yes, need to enable that per session & globally.

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs (hunk 3 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs]: If you call `Delete(string id)`, that is a good indication that you don't care what the current value is.

**Comment by ppekrol** [comment]
Rebased

**Review by karmeli87 — COMMENTED** [review]
LGTM, but what will happened if...

1. mixing cluster-tx with normal I think it shouldn't be an issue.. IIRC we always prefer the document of the cluster-tx
2. conflicts or updates because of an external replication... we are replicating the document but not the compare exchange values.. which might be a problem?

**Comment by ayende** [comment]
The document contains a `@cluster-tx-id` field, which is set to the index of the compare exchange value.
In such a case, we'll have a problem, yes. I'm not sure what would be a good way to resolve it.

If the field is now out of sync of the compare exchange, what should we do?

For that matter, we probably need to strip this value on external replication, because that is a per cluster setting.

We can try to do a repair on the cluster transaction value, I guess? But I'm not sure how viable that would be.

**Comment by karmeli87** [comment]
Maybe we can add the TopologyId to the compare-exchange name and to the metadata to make it some sort of multi-cluster change vector.
`@multi-cluster-change-vector`:`<topology1>:<raft-index>, <topology2>:<raft-index>`

One issue that I can think with this approach is in import to new database, I think.

**Comment by ayende** [comment]
That wouldn't help the scenario of misaligned updates, thoughs.

**Comment by ayende** [comment]
Consider this scenario:

* I load a document with `cluster-tx-id` of 7 in tx A.
* Modify that in a cluster tx B.
* The compare exchange value is now 8.
* I call `SaveChanges()` on tx A, overwriting the old value and returning the *`cluster-tx-id` to be. 
* All cluster wide operations will fail now on that document.

**Comment by ryanheath** [comment]
Nice to see this is coming to RavenDB, making cluster wide transactions a whole lot easier to work with!
Mad props @ayende !

File: test/SlowTests/Issues/RavenDB-16614.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+            }
+
+            using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
+            {
+                var user = await session.LoadAsync<User>("users/arava");
+                user.Name = "Arava Eini";
+                var user2 = await session.LoadAsync<User>("users/phoebe");
+                user2.Name = "Phoebe Eini";
+
+                using (var conflictedSession = store.OpenAsyncSession(new SessionOptions {TransactionMode = TransactionMode.ClusterWide}))
```
Comment (ryanheath) [inline test/SlowTests/Issues/RavenDB-16614.cs:286]: How can we make sure that this session is hitting another node? Or is this done implicit by the test runner?
If implicit, how can we make sure the test also succeeds with hitting one node each time?

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs [hunk 5]
```diff
@@ -190,15 +200,84 @@ public List<string> ExecuteCompareExchangeCommands(ClusterOperationContext conte
...
+
         public unsafe void SaveCommandsBatch(ClusterOperationContext context, long index)
         {
             if (HasDocumentsInTransaction == false)
                 return;
 
             var items = context.Transaction.InnerTransaction.OpenTable(ClusterStateMachine.TransactionCommandsSchema, ClusterStateMachine.TransactionCommands);
             var commandsCountPerDatabase = context.Transaction.InnerTransaction.ReadTree(ClusterStateMachine.TransactionCommandsCountPerDatabase);
-            var commands = SerializedDatabaseCommands.Clone(context);
+            var commands = context.ReadObject(SerializedDatabaseCommands, "serialized-tx-commands");
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:279]: Is this change needed?

Comment (ayende) [reply src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:279]: Yes, we are mutating the `SerializedDatabaseCommands` to add artifical commands here.

File: test/SlowTests/Issues/RavenDB-16614.cs (hunk 4 above)
Comment (ayende) [reply test/SlowTests/Issues/RavenDB-16614.cs:286]: We aren't trying to do that. Cluster Wide transactions are going to go to the leader anyway, so it doesn't matter from our perspective.

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs [hunk 6]
```diff
@@ -58,6 +58,8 @@ public abstract partial class InMemoryDocumentSessionOperations : IDisposable
 
         private BatchOptions _saveChangesOptions;
 
+        public bool DisableAtomicDocumentWritesInClusterWideTransaction;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs]: internal readonly?

File: src/Raven.Server/Documents/TcpHandlers/TcpConnectionOptions.cs [hunk 7]
```diff
@@ -114,9 +114,10 @@ public void Dispose()
         }
 
 #if !RELEASE
+        private string ss = Environment.StackTrace;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TcpHandlers/TcpConnectionOptions.cs]: Can we remove this?

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs [hunk 8]
```diff
@@ -69,6 +70,7 @@ public class ClusterTransactionOptions : IDynamicJson
             public TimeSpan? WaitForIndexesTimeout;
             public bool WaitForIndexThrow;
             public List<string> SpecifiedIndexesQueryString;
+            public bool DisableAtomicDocumentWrites;
```
Comment (ppekrol) [inline src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:79]: This will be false for legacy commands? Is this OK?

**Review by ppekrol — COMMENTED** [review]
@ayende can you check failing tests?

File: src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs (hunk 6 above)
Comment (ayende) [reply src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs]: No, this is so the user can disable that for a single session.

File: src/Raven.Server/Documents/TcpHandlers/TcpConnectionOptions.cs (hunk 7 above)
Comment (ayende) [reply src/Raven.Server/Documents/TcpHandlers/TcpConnectionOptions.cs]: Yes, I added this to see if it would fail with additional information.

File: src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs (hunk 8 above)
Comment (ayende) [reply src/Raven.Server/ServerWide/Commands/ClusterTransactionCommand.cs:79]: I don't think that this is a scenario that we need to worry about.

**Comment by ayende** [comment]
Here is what the document will look like:

```
{
    "Name": "arava",
    "@metadata": {
        "@collection": "Users",
        "Raven-Clr-Type": "SlowTests.Issues.RavenDB_16614+User, SlowTests",
        "@flags": "FromClusterTransaction",
        "@cluster-transaction-index": {
            "Au3CeLCT8kWAU1UB4QcDuQ": 11
        }
    }
}
```

File: src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs [hunk 9]
```diff
@@ -15,10 +15,21 @@ namespace Raven.Client.Documents.Commands.Batches
...
         {
+            DisableAtomicDocumentWrites = disableAtomicDocumentsWrites;
+        }
+
+        protected override void AppendOptions(StringBuilder sb)
+        {
+            base.AppendOptions(sb);
+            if (DisableAtomicDocumentWrites)
+            {
+                sb.Append("?disableAtomicDocumentWrites=true");
```
Comment (garayx) [inline src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs]: should be `&disableAtomicDocumentWrites=true` if ` _options` are not null?

Comment (ayende) [reply src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs]: I don't follow?

Comment (ppekrol) [reply src/Raven.Client/Documents/Commands/Batches/BatchCommand.cs]: At the end if null then we should not send anything (and then use what s configured in the database), if not null then send whatever is configured because client will override it.

**Comment by ayende** [comment]
Adding the newest commits just to have a place for them. I'll do a clean PR tomoor.w

**Comment by ayende** [comment]
Closing in favor of a cleanup PR
